pallets/xvm/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+// This file is part of Astar.
+
+// Copyright (C) 2019-2023 Stake Technologies Pte.Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+//! # XVM pallet
+//!
+//! A module to provide
+//!
+//! ## Overview
+//!
+//! The XVM pallet provides a runtime interface to call different VMs. It currently
+//! supports two VMs: EVM and WASM. With further development, more VMs can be added.
+//!
+//! Together with other functionalities like Chain Extension and precompiles,
+//! the XVM pallet enables the runtime to support cross-VM calls.
+//!
+//! ## Interface
+//!
+//! ### Implementation
+//!
+//! - Implements `XvmCall` trait.
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{ensure, traits::ConstU32, BoundedVec};
+use pallet_contracts::{CollectEvents, DebugInfo, Determinism};
+use pallet_evm::GasWeightMapping;
+use parity_scale_codec::Decode;
+use sp_core::U256;
+use sp_runtime::traits::StaticLookup;
+use sp_std::{marker::PhantomData, prelude::*};
+
+use astar_primitives::{
+    ethereum_checked::{
+        AccountMapping, CheckedEthereumTransact, CheckedEthereumTx, MAX_ETHEREUM_TX_INPUT_SIZE,
+    },
+    xvm::{CallError, CallErrorWithWeight, CallInfo, CallResult, Context, VmId, XvmCall},
+};
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+pub use weights::WeightInfo;
+
+#[cfg(test)]
+mod mock;
+
+pub use pallet::*;
+
+pub type WeightInfoOf<T> = <T as Config>::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_contracts::Config {
+        /// Mapping from `Account` to `H160`.
+        type AccountMapping: AccountMapping<Self::AccountId>;
+
+        /// Mapping from Ethereum gas to Substrate weight.
+        type GasWeightMapping: GasWeightMapping;
+
+        /// `CheckedEthereumTransact` implementation.
+        type EthereumTransact: CheckedEthereumTransact;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+}
+
+impl<T: Config> XvmCall<T::AccountId> for Pallet<T> {
+    fn call(
+        context: Context,
```

Comment:
Thinking about this, both EVM & Parity's WASM allow limiting of how much gas can be spent by the cross-contract call. We should also support this, I guess?

Not necessarily for this PR, but would be good to do in a follow-up.

Reply:
Created an issue for this. Will follow up. https://github.com/AstarNetwork/Astar/issues/987